File: packages/connectivity_status/ios/connectivity_status/Sources/connectivity_status/ConnectivityStatusPlugin.swift
import Flutter
import Network
#if os(iOS)
import UIKit
#endif

/// Exposes the OS metered capability (`NWPath.isExpensive`, which reflects
/// the underlying transport even when a VPN masks it) over the
/// `net.kaiwu.connectivity_status/metered` MethodChannel.
///
/// Registered automatically by `GeneratedPluginRegistrant` — no manual
/// call needed in a consuming app's `AppDelegate`. Each query spins up a
/// fresh `NWPathMonitor` and reads the current path once, so the answer can
/// never be staler than the connection-type list it is joined with on the
/// Dart side.
public class ConnectivityStatusPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "net.kaiwu.connectivity_status/metered",
      binaryMessenger: registrar.messenger()
    )
    let instance = ConnectivityStatusPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard call.method == "isActiveNetworkMetered" else {
      result(FlutterMethodNotImplemented)
      return
    }
    Self.readMetered(result: result)
  }

  /// Reads the current network path's metered flag via a one-shot
  /// `NWPathMonitor`, then tears it down.
  private static func readMetered(result: @escaping FlutterResult) {
    let monitor = NWPathMonitor()
    let queue = DispatchQueue(label: "net.kaiwu.connectivity_status.metered")
    var replied = false

    monitor.pathUpdateHandler = { path in
      // Single-reply guard: take the first delivered path, then tear the
      // monitor down. Cancelling and niling the handler breaks the
      // monitor↔closure retain cycle so this per-call monitor deallocates.
      guard !replied else {
        return
      }
      replied = true

      let metered = path.isExpensive
      monitor.cancel()
      monitor.pathUpdateHandler = nil

      // FlutterResult must be delivered on the platform (main) thread.
      DispatchQueue.main.async {
        result(metered)
      }
    }

    monitor.start(queue: queue)
  }
}

File: packages/connectivity_status/android/src/main/kotlin/net/kaiwu/connectivity_status/ConnectivityStatusPlugin.kt
package net.kaiwu.connectivity_status

import android.content.Context
import android.net.ConnectivityManager
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** Exposes the OS metered capability
 *  (`ConnectivityManager.isActiveNetworkMetered`, which reflects the
 *  underlying transport even when a VPN masks it) over the
 *  `net.kaiwu.connectivity_status/metered` MethodChannel.
 *
 *  Requires `android.permission.ACCESS_NETWORK_STATE`. Registered
 *  automatically by `GeneratedPluginRegistrant` — no manual call needed in
 *  a consuming app's `MainActivity`. */
class ConnectivityStatusPlugin :
    FlutterPlugin,
    MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var context: Context

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        channel = MethodChannel(
            flutterPluginBinding.binaryMessenger,
            "net.kaiwu.connectivity_status/metered",
        )
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(
        call: MethodCall,
        result: Result,
    ) {
        if (call.method != "isActiveNetworkMetered") {
            result.notImplemented()
            return
        }
        val manager = context.getSystemService(Context.CONNECTIVITY_SERVICE)
            as? ConnectivityManager
        if (manager == null) {
            result.error("METERED_QUERY_FAILED", "metered query failed", null)
            return
        }
        try {
            result.success(manager.isActiveNetworkMetered)
        } catch (e: SecurityException) {
            // Missing ACCESS_NETWORK_STATE → degrade to the Dart-side
            // heuristic. Any other exception is a genuine bug and is left
            // to propagate (no opaque catch-all masking it).
            result.error("METERED_QUERY_FAILED", "metered query failed", null)
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}
